File: algorithm/Python3_version/leetcode1053.previous-permutation-with-one-swap.py
class Solution:
    def prevPermOpt1(self, A: List[int]) -> List[int]:
        l, r = 0, 0
        for i in range(1,len(A)):
            if A[i-1] > A[i]:
                l, r = i-1, i
            elif A[i] < A[l] and A[i] > A[r]:
                r = i
        t = A[l]
        A[l] = A[r]
        A[r] = t
        return A

File: algorithm/leetcode1053.previous-permutation-with-one-swap.cpp
class Solution {
public:
    vector<int> prevPermOpt1(vector<int>& A) {
        int i = 0, l = 0, r = 0;
        for(i = 1; i < A.size(); ++i)
        {
            if(A[i-1] > A[i])//遇到下降
                l = i-1, r = i;
            //后序找到小于 A[i-1] 的数
            else if(A[i] < A[l] && A[i] > A[r])
                r = i;
        }
        swap(A[l], A[r]);
        return A;
    }
};